src/hotspot/share/gc/shared/stringdedup/stringDedup.hpp

```diff
@@ -25,108 +25,186 @@
-// used by String objects. When deduplicating, a lookup is made in this table to see
-// if there is already an identical character array somewhere on the heap. If so, the
-// String object is adjusted to point to that character array, releasing the reference
-// to the original array allowing it to eventually be garbage collected. If the lookup
-// fails the character array is instead inserted into the hashtable so that this array
-// can be shared at some point in the future.
+// The set of requests uses entries from a pair of weak OopStorage objects.
+// One is used for requests, the other is being processed.  When processing
+// completes, the roles of the storage objects are exchanged.  The GC adds
+// entries referring to discovered candidates, allocating new OopStorage
+// entries for the requests.  The deduplication processing thread does a
+// concurrent iteration over the processing storage, deduplicating the
+// Strings and releasing the OopStorage entries.  Two storage objects are
+// used so there isn't any conflict between adding and removing entries by
+// different threads.
 //
-// Candidate selection criteria is GC specific.
+// The table uses entries from another weak OopStorage to hold the byte
+// arrays.  This permits reclamation of arrays that become unused.  This is
+// separate from the request storage objects because dead count tracking is
+// used by the table implementation as part of resizing decisions and for
+// deciding when to cleanup dead entries in the table.  The usage pattern
+// for the table is also very different from that of the request storages.
+// The request/processing storages are used in a way that supports bulk
+// allocation and release of entries.
 //
-// Interned strings are a bit special. They are explicitly deduplicated just before
-// being inserted into the StringTable (to avoid counteracting C2 optimizations done
-// on string literals), then they also become deduplication candidates if they reach
-// the deduplication age threshold or are evacuated to an old heap region. The second
-// attempt to deduplicate such strings will be in vain, but we have no fast way of
-// filtering them out. This has not shown to be a problem, as the number of interned
-// strings is usually dwarfed by the number of normal (non-interned) strings.
+// Candidate selection criteria is GC specific.  This class provides some
+// helper functions that may be of use when implementing candidate
+// selection.
+//
+// Strings interned in the StringTable require special handling.  Once a
+// String has been added to the StringTable, its byte array must not change.
+// Doing so would counteract C2 optimizations on string literals.  But an
+// interned string might later become a deduplication candidate through the
+// normal GC discovery mechanism.  This is handled by setting the
+// no_deduplication flag in a string before interning it.  A string with
```

Comment:
s/string/String

Maybe:

s/This is handled.../Deduplication is prevented by setting the no_deduplication flag in the String before.../